### PR TITLE
Improve docstrings and add function support to test

### PR DESCRIPTION
**Pull request by LilianBoulard**
- Slightly improves the style of numpydoc-valid methods
- Adds the support of public-facing functions in `test_docstrings.py`

### CODE REVIEW

File: dirty_cat/_datetime_encoder.py [hunk 1]
```diff
@@ -64,10 +64,10 @@ class DatetimeEncoder(BaseEstimator, TransformerMixin):
         Number of features in the data seen during fit.
     n_features_out_: int
         Number of features of the transformed data.
-    features_per_column_: typing.Dict[int, typing.List[str]]
+    features_per_column_: mapping of int to list of str
```
Comment (GaelVaroquaux) [inline dirty_cat/_datetime_encoder.py:67]: I have the impression that we are always changing those typing documentation. This is making review hard, and I fail to see where the benefit is for the user.

File: dirty_cat/tests/test_docstrings.py [hunk 2]
```diff
@@ -77,8 +74,16 @@ def get_all_methods():
             yield Estimator, method
 
 
+def get_functions_to_validate():
```
Comment (GaelVaroquaux) [inline dirty_cat/tests/test_docstrings.py:77]: What's the difference between this function and `get_public_functions` ?

**Review by GaelVaroquaux — COMMENTED** [review]
I think that I reviewed this right, though I worry about reviewing such a PR with many scattered changes.

One comment, though

File: dirty_cat/tests/test_docstrings.py (hunk 2 above)
Comment (LilianBoulard) [reply dirty_cat/tests/test_docstrings.py:77]: It's an intermediary function to `get_public_functions` used by by the parametrization, line 194, just like `get_methods_to_validate` is for `get_public_classes`.
I'm just keeping the process similar for both public functions and class methods.

File: dirty_cat/_fuzzy_join.py [hunk 3]
```diff
@@ -42,39 +42,39 @@ def fuzzy_join(
...
     Parameters
     ----------
-    left : pandas.DataFrame
+    left : :class:`~pandas.DataFrame`
         A table to merge.
-    right : pandas.DataFrame
+    right : :class:`~pandas.DataFrame`
         A table used to merge with.
-    how: typing.Literal["left", "right"], default=`left`
+    how: {"left", "right"}, optional, default=`left`
```
Comment (jovan-stojanovic) [inline dirty_cat/_fuzzy_join.py:49]:
```suggestion
    how : {"left", "right"}, optional, default=`left`
```

File: dirty_cat/_fuzzy_join.py [hunk 4]
```diff
@@ -85,12 +85,12 @@ def fuzzy_join(
...
-    suffixes : typing.Tuple[str, str], default=('_x', '_y')
+    suffixes : str 2-tuple, default=('_x', '_y')
         A list of strings indicating the suffix to add when overlaping
         column names.
 
-    Returns:
-    --------
+    Returns
+    -------
     df_joined: pandas.DataFrame
```
Comment (jovan-stojanovic) [inline dirty_cat/_fuzzy_join.py:94]:
```suggestion
    df_joined : :class:`~pandas.DataFrame`
```

**Review by jovan-stojanovic — COMMENTED** [review]
Some small changes before merging

File: dirty_cat/_datetime_encoder.py [hunk 5]
```diff
@@ -64,10 +64,10 @@ class DatetimeEncoder(BaseEstimator, TransformerMixin):
         Number of features in the data seen during fit.
     n_features_out_: int
```
Comment (jovan-stojanovic) [inline dirty_cat/_datetime_encoder.py:65]:
```suggestion
    n_features_out_ : int
```

File: dirty_cat/_datetime_encoder.py (hunk 1 above)
Comment (jovan-stojanovic) [inline dirty_cat/_datetime_encoder.py:67]:
```suggestion
    features_per_column_ : mapping of int to list of str
```

File: dirty_cat/_datetime_encoder.py [hunk 6]
```diff
@@ -64,10 +64,10 @@ class DatetimeEncoder(BaseEstimator, TransformerMixin):
...
         Number of features of the transformed data.
-    features_per_column_: typing.Dict[int, typing.List[str]]
+    features_per_column_: mapping of int to list of str
         Dictionary mapping the index of the original columns
         to the list of features extracted for each column.
-    col_names_: typing.Optional[typing.List[str]]
+    col_names_: None or list of str
```
Comment (jovan-stojanovic) [inline dirty_cat/_datetime_encoder.py:70]:
```suggestion
    col_names_ : None or list of str
```

**Review by jovan-stojanovic — COMMENTED** [review]
And one more..

**Comment by LilianBoulard** [comment]
Thanks for the fixes!
The cleanup of `fuzzy_join` deserves its own PR (see #345), so I'll merge it like that!

**Pull request merged**